File: problem-solving/baekjoon/2164_카드2.go
// https://www.acmicpc.net/problem/2164
// 큐를 사용하여 동작을 구현하는 문제
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n int
	fmt.Fscanln(reader, &n)

	var queue []int
	for i := 1; i <= n; i++ {
		queue = append(queue, i)
	}

	var discard bool = true
	for len(queue) > 1 {
		if discard {
			queue = queue[1:]
		} else {
			queue = append(queue[1:], queue[0])
		}
		discard = !discard
	}

	fmt.Fprintln(writer, queue[0])
}

File: problem-solving/baekjoon/2164_카드2.py
# https://www.acmicpc.net/problem/2164
# 큐를 사용하여 동작을 구현하는 문제
import sys
from collections import deque

if __name__ == "__main__":
    n = int(sys.stdin.readline())
    q = deque(maxlen=n)

    for i in range(n):
        q.append(i+1)
    
    discard = True
    while len(q) > 1:
        if discard:
            q.popleft()
        else:
            q.append(q.popleft())
        discard = not discard
    
    print(q.pop())